services/github/github-constellation.js

```diff
@@ -5,6 +5,11 @@ import GithubApiProvider from './github-api-provider.js'
 import { setRoutes as setAdminRoutes } from './auth/admin.js'
 import { setRoutes as setAcceptorRoutes } from './auth/acceptor.js'
 
+const readPackagesScope = 'read:packages'
+// Multiple scopes need to be uri-encoded space delimited
+const tokenScopes = `${readPackagesScope}`
```

Comment:
I think even if we don't have any integrations that use them yet, it probably makes sense to start _requesting_ more scopes sooner rather than later. i.e: even if we only add `packageScopedTokens` to `GithubApiProvider` now, if we think we've got a good case for needing `public_repo`, `read:user`, `read:org`, etc. I think it makes sense to have a think and make some decisions on that now and implement stuff which needs them later. The reason I say this is because the process of trying to convince new users to donate a token or existing users to re-authorize is in many ways the most cumbersome part of this process and GitHub does not provide any nice way of doing this.. or at least it is the part of the process that is an exercise in cat herding. We have to convince people to do a thing for us - we can't solve it with code, even if its really good code. There's also an element of user 'fatigue' to consider here. If we ask users for 3 new scopes once I suspect we'll get more total takeup than asking for one new scope 3 times over the next 6 months for example.

I think its inevitable that we will have to go through that process more than once over time, but if we can think about it a bit now and reduce the total number of times we need to do it, that is going to be very advantageous.

Just to start collating our pile of possible scopes we might (or might not) want to ask for aside from `package:read`, I think the others I know of are:

- `public_repo`
- `read:user`
- `read:org`

Are there any scopes you know of that would be needed for any other open feature requests that are blocked on this?

Reply:
> I think even if we don't have any integrations that use them yet, it probably makes sense to start requesting more scopes sooner rather than later

Agreed. We do know that it's impossible to completely future-proof ourselves from _ever_ having to go through this again, but we should absolutely strive to cover everything known any time we need to request more scopes, and I'd like for us to identify the full set of scopes we want for this pass before merging the PR.

The balance of course is not overbidding on scopes in a way that will turn people off/reduce the population of those willing to donate. For example, I've paused/avoided authorizing apps that request the `read:user` scope because it grants full access to all private profile information (including contact info).

Obviously we store donated tokens securely and have no intention of ever using them for anything beyond GitHub API calls to fetch data needed for badges. However, I do think it's important to think about it from the concerns/perspective of users considering donating/granting scope increases.

> Are there any scopes you know of that would be needed for any other open feature requests that are blocked on this?

The only scenario I recall that also hit the scope wall was #6104, which would've needed `public_repo` for repository-level projects, and `read:org` for organization level projects. I don't think that first one is going to be viable regardless, as `public_repo` also includes write permissions (https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps)

> Limits access to public repositories. **That includes read/write access to code**, commit statuses, repository projects, collaborators, and deployment statuses for public repositories and organizations. Also required for starring public repositories.

I personally wouldn't authorize a token with that level of scope to anyone, so I wouldn't feel right asking Shields users to do this. In my opinion that feels like a use case best handled via a custom endpoint badge where users could individually wire up their endpoint with an appropriately scoped token they keep/maintain.

`read:org` _might_ be more viable, though it's unclear to me whether that's limited to just exposing the organization/team mapping for that specific user (i.e. what organizations and what org teams is the owner of the token a member of), or if it can be used to see _all_ org and org team membership for any org/team the token owner is a member of. The former would be okay, but the latter would give me pause to say the least as I believe GitHub allows users to choose whether to convey their membership publicly for a reason and such a token would allow token-donators to theoretically expose that info of _other_ org/team members without their consent.

At least at this point in time, I'd lean towards directing all GitHub Project-related use cases to the Custom Endpoint path, and we could even consider implementing an example endpoint and linking out to it (e.g. from an issue/discussion in this repo and/or our awesome list)

Reply:
Yes you're right. We shouldn't ask for public_repo (or any other scope that grants write access). Good shout.